### PR TITLE
expression: implement vectorized evaluation for `builtinGetFormatSig`

### PR DESCRIPTION
**Pull request by b41sh**
### What problem does this PR solve? 
Implement vectorized evaluation for `builtinGetFormatSig`.
[12106](https://github.com/pingcap/tidb/issues/12106)

### What is changed and how it works?
```
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinTimeFunc/builtinGetFormatSig-VecBuiltinFunc-4       28806   41637 ns/op    0 B/op    0 allocs/op
BenchmarkVectorizedBuiltinTimeFunc/builtinGetFormatSig-NonVecBuiltinFunc-4    15163    80904 ns/op    0 B/op    0 allocs/op
```

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: expression/builtin_time_vec.go [hunk 1]
```diff
@@ -435,11 +435,84 @@ func (b *builtinNullTimeDiffSig) vecEvalDuration(input *chunk.Chunk, result *chu
...
+			case jisLocation:
+				res = "%H:%i:%s"
+			case isoLocation:
+				res = "%H:%i:%s"
+			case eurLocation:
+				res = "%H.%i.%s"
+			case internalLocation:
+				res = "%H%i%s"
+			}
+		}
```
Comment (qw4990) [inline expression/builtin_time_vec.go]: How about creating a new function for these lines like `func (b *builtinGetFormatSig) getFormat(format, location string) string`?

**Comment by qw4990** [comment]
And please resolve these conflicts. @b41sh

File: expression/bench_test.go [hunk 2]
```diff
@@ -523,6 +523,33 @@ func (g *randDurInt) gen() interface{} {
...
+		usaLocation,
+		jisLocation,
+		isoLocation,
+		eurLocation,
+		internalLocation,
+	}
+	return locations[rand.Intn(5)]
+}
+
+// formatGener is used to generate format
```
Comment (XuHuaiyu) [inline expression/bench_test.go]: is used to generate a format for the built-in function `GetFormat`.

Comment (b41sh) [reply expression/bench_test.go]: ok

File: expression/builtin_time_vec.go (hunk 1 above)
Comment (b41sh) [reply expression/builtin_time_vec.go]: ok

File: expression/bench_test.go (hunk 2 above)
Comment (b41sh) [reply expression/bench_test.go]: we can use `timeFormatGener`

Comment (XuHuaiyu) [reply expression/bench_test.go]: yep

File: expression/builtin_time_vec.go [hunk 3]
```diff
@@ -470,11 +470,88 @@ func (b *builtinNullTimeDiffSig) vecEvalDuration(input *chunk.Chunk, result *chu
...
+		format := buf0.GetString(i)
+		location := buf1.GetString(i)
+		res := b.getFormat(format, location)
+		result.AppendString(res)
+	}
+	return nil
+}
+
+func (b *builtinGetFormatSig) getFormat(format, location string) string {
```
Comment (qw4990) [inline expression/builtin_time_vec.go]: Please make the row-oriented implementation also use this function.

Comment (qw4990) [reply expression/builtin_time_vec.go]: Rest, LGTM

Comment (qw4990) [reply expression/builtin_time_vec.go]: And please resolve conflicts, PTAL ~ @b41sh

Comment (b41sh) [reply expression/builtin_time_vec.go]: ok

File: expression/bench_test.go [hunk 4]
```diff
@@ -541,6 +541,54 @@ func (g *randDurInt) gen() interface{} {
...
+// locationGener is used to generate location for the built-in function GetFormat.
+type locationGener struct {
+	nullRation float64
+}
+
+func (g *locationGener) gen() interface{} {
+	if rand.Float64() < g.nullRation {
+		return nil
+	}
+	switch rand.Uint32() % 4 {
```
Comment (Reminiscent) [inline expression/bench_test.go]: Here should be ```switch rand.Uint32() % 5 {```

File: expression/bench_test.go [hunk 5]
```diff
@@ -541,6 +541,54 @@ func (g *randDurInt) gen() interface{} {
...
+// formatGener is used to generate a format for the built-in function GetFormat.
+type formatGener struct {
+	nullRation float64
+}
+
+func (g *formatGener) gen() interface{} {
+	if rand.Float64() < g.nullRation {
+		return nil
+	}
+	switch rand.Uint32() % 3 {
```
Comment (Reminiscent) [inline expression/bench_test.go]: Here should be ```rand.Uint32() % 4```.

File: expression/bench_test.go (hunk 4 above)
Comment (b41sh) [reply expression/bench_test.go]: ok

**Pull request merged**